File: rust/src/sync.rs
use anyhow::{Context as _, Result};
use std::fs;
use std::path::Path;

use crate::store::{ContextStore, PeerConfig};

pub struct SyncResult {
    pub peer_name: String,
    pub pulled: Vec<String>,
    pub pushed: Vec<String>,
    pub errors: Vec<String>,
}

// Two transports: HTTP for WAN (daemon serves context over the internet) and
// SSH/rsync for LAN (leverages ~/.ssh/config for auth, no daemon needed).
// SSH is preferred for trusted home-lab peers; HTTP for public registry peers.
enum Transport {
    Http { base_url: String },
    Ssh { host: String, path: String },
}

fn parse_url(url: &str) -> Result<Transport> {
    if url.starts_with("http://") || url.starts_with("https://") {
        Ok(Transport::Http {
            base_url: url.trim_end_matches('/').to_string(),
        })
    } else if url.starts_with("ssh://") {
        // ssh://hostname:/path or ssh://user@hostname:/path
        // hostname can be an SSH config alias — we pass it straight to rsync/ssh
        let rest = &url[6..];
        let (host, path) = rest
            .split_once(':')
            .context("SSH URL must be ssh://host:/path")?;
        // Prevent argument injection: rsync treats leading '-' as flags, and shell
        // metacharacters in the host could execute arbitrary commands. A malicious
        // peer URL like "ssh://--server:/" would become `rsync -az --server:/ ...`.
        if host.starts_with('-') || path.starts_with('-') {
            anyhow::bail!("Invalid SSH URL: host/path cannot start with '-'");
        }
        if host.contains(';') || host.contains('|') || host.contains('`') || host.contains('$') {
            anyhow::bail!("Invalid SSH URL: host contains shell metacharacters");
        }
        Ok(Transport::Ssh {
            host: host.to_string(),
            path: path.to_string(),
        })
    } else {
        anyhow::bail!("Unknown URL scheme: {url}. Use http:// or ssh://")
    }
}

pub async fn sync_all(store: &ContextStore) -> Result<Vec<SyncResult>> {
    let config = store.read_config()?;
    let mut results = vec![];

    for peer in &config.peers {
        match sync_peer(store, peer).await {
            Ok(result) => results.push(result),
            Err(e) => results.push(SyncResult {
                peer_name: peer.name.clone(),
                pulled: vec![],
                pushed: vec![],
                errors: vec![format!("sync failed: {}", e)],
            }),
        }
    }

    Ok(results)
}

pub async fn sync_one(store: &ContextStore, peer_name: &str) -> Result<SyncResult> {
    let config = store.read_config()?;
    let peer = config
        .peers
        .iter()
        .find(|p| p.name == peer_name || p.id == peer_name)
        .context(format!("Peer not found: {}", peer_name))?
        .clone();
    sync_peer(store, &peer).await
}

async fn sync_peer(store: &ContextStore, peer: &PeerConfig) -> Result<SyncResult> {
    let transport = parse_url(&peer.url)?;
    let peer_dir = store.root().join(".peers").join(&peer.name);
    fs::create_dir_all(&peer_dir)?;

    match transport {
        Transport::Http { base_url } => sync_http(store, peer, &base_url, &peer_dir).await,
        Transport::Ssh { host, path } => sync_ssh(store, peer, &host, &path, &peer_dir).await,
    }
}

// --- HTTP sync (WAN) ---

async fn sync_http(
    store: &ContextStore,
    peer: &PeerConfig,
    base_url: &str,
    peer_dir: &Path,
) -> Result<SyncResult> {
    let client = reqwest::Client::new();
    let mut pulled = vec![];
    let mut pushed = vec![];
    let mut errors = vec![];

    // Pull root files
    for file in &["CONTEXT.md", "PROFILE.md"] {
        match client
            .get(format!("{}/read/{}", base_url, file))
            .send()
            .await
        {
            Ok(resp) if resp.status().is_success() => {
                let body = resp.bytes().await?;
                fs::write(peer_dir.join(file), &body)?;
                pulled.push(file.to_string());
            }
            Ok(resp) => errors.push(format!("{}: HTTP {}", file, resp.status())),
            Err(e) => errors.push(format!("{}: {}", file, e)),
        }
    }

    // Pull directories (shared/, knowledge/)
    for dir in &["shared", "knowledge"] {
        match pull_http_dir(&client, base_url, dir, peer_dir).await {
            Ok(files) => pulled.extend(files),
            Err(e) => errors.push(format!("{}/: {}", dir, e)),
        }
    }

    // Push outbox → peer inbox
    match push_http_outbox(store, &client, base_url, peer).await {
        Ok(files) => pushed.extend(files),
        Err(e) => errors.push(format!("push: {}", e)),
    }

    fs::write(
        peer_dir.join(".last-sync"),
        chrono::Utc::now().to_rfc3339(),
    )?;

    Ok(SyncResult {
        peer_name: peer.name.clone(),
        pulled,
        pushed,
        errors,
    })
}

#[derive(serde::Deserialize)]
struct FileEntry {
    name: String,
    is_dir: bool,
    #[allow(dead_code)]
    size: u64,
}

async fn pull_http_dir(
    client: &reqwest::Client,
    base_url: &str,
    dir: &str,
    peer_dir: &Path,
) -> Result<Vec<String>> {
    let mut pulled = vec![];
    let resp = client
        .get(format!("{}/ls/{}", base_url, dir))
        .send()
        .await?;

    if !resp.status().is_success() {
        return Ok(pulled);
    }

    let files: Vec<FileEntry> = resp.json().await?;
    let local_dir = peer_dir.join(dir);
    fs::create_dir_all(&local_dir)?;

    for f in files {
        if f.is_dir {
            continue;
        }
        // Remote peers control the filename in the /ls response — a malicious peer
        // could return "../../../etc/cron.d/backdoor". Extract basename to contain writes.
        let safe_name = match std::path::Path::new(&f.name)
            .file_name()
            .and_then(|n| n.to_str())
            .filter(|n| !n.is_empty() && !n.contains(".."))
        {
            Some(n) => n.to_string(),
            None => continue, // skip malicious filenames
        };
        let resp = client
            .get(format!("{}/read/{}/{}", base_url, dir, safe_name))
            .send()
            .await?;
        if resp.status().is_success() {
            let body = resp.bytes().await?;
            fs::write(local_dir.join(&safe_name), &body)?;
            pulled.push(format!("{}/{}", dir, safe_name));
        }
    }

    Ok(pulled)
}

async fn push_http_outbox(
    store: &ContextStore,
    client: &reqwest::Client,
    base_url: &str,
    peer: &PeerConfig,
) -> Result<Vec<String>> {
    let mut pushed = vec![];
    let outbox_dir = store.root().join("outbox");
    if !outbox_dir.exists() {
        return Ok(pushed);
    }

    for entry in fs::read_dir(&outbox_dir)? {
        let entry = entry?;
        let fname = entry.file_name().to_string_lossy().to_string();
        if !fname.ends_with(".json") {
            continue;
        }
        // Only push messages addressed to this peer
        if !fname.contains(&format!("_to-{}", peer.name)) && !fname.contains(&peer.id) {
            continue;
        }

        let body = fs::read(entry.path())?;
        match client
            .post(format!("{}/inbox", base_url))
            .header("Content-Type", "application/json")
            .body(body)
            .send()
            .await
        {
            Ok(r) if r.status().is_success() => {
                // Move to .sent/ to prevent re-delivery
                archive_sent(&outbox_dir, &fname)?;
                pushed.push(fname);
            }
            Ok(r) => anyhow::bail!("{}: HTTP {}", fname, r.status()),
            Err(e) => anyhow::bail!("{}: {}", fname, e),
        }
    }

    Ok(pushed)
}

// --- SSH sync (LAN) ---
// Uses rsync over SSH. Host can be an SSH config alias (e.g. "alice.local")
// so we get hostname resolution, user, and key from ~/.ssh/config for free.

async fn sync_ssh(
    store: &ContextStore,
    peer: &PeerConfig,
    host: &str,
    remote_path: &str,
    peer_dir: &Path,
) -> Result<SyncResult> {
    let mut pulled = vec![];
    let mut pushed = vec![];
    let mut errors = vec![];

    // Pull root files
    for file in &["CONTEXT.md", "PROFILE.md"] {
        let src = format!("{}:{}/{}", host, remote_path, file);
        let dst = peer_dir.join(file);
        match rsync(&src, &dst.to_string_lossy()).await {
            Ok(_) => pulled.push(file.to_string()),
            Err(e) => errors.push(format!("{}: {}", file, e)),
        }
    }

    // Pull directories
    for dir in &["shared", "knowledge"] {
        let local = peer_dir.join(dir);
        fs::create_dir_all(&local)?;
        let src = format!("{}:{}/{}/", host, remote_path, dir);
        let dst = format!("{}/", local.to_string_lossy());
        match rsync_dir(&src, &dst).await {
            Ok(_) => pulled.push(format!("{}/", dir)),
            Err(e) => errors.push(format!("{}/: {}", dir, e)),
        }
    }

    // Pull peer's outbox for messages addressed to us — peer may be behind NAT
    // and can't push to us, so we grab messages they left for us.
    let config = store.read_config()?;
    let my_name = &config.name;
    let inbox_dir = store.root().join("inbox");
    fs::create_dir_all(&inbox_dir)?;
    {
        let src = format!("{}:{}/outbox/", host, remote_path);
        let dst = format!("{}/", inbox_dir.to_string_lossy());
        let output = tokio::process::Command::new("rsync")
            .args([
                "-az", "--ignore-existing",
                "--include", &format!("*_to-{}.json", my_name),
                "--include", "*_to-all.json",
                "--exclude", "*",
                &src, &dst,
            ])
            .output()
            .await;
        match output {
            Ok(o) if o.status.success() => { pulled.push("outbox→inbox".to_string()); }
            Ok(o) => {
                let stderr = String::from_utf8_lossy(&o.stderr);
                if !stderr.contains("No such file") {
                    errors.push(format!("pull outbox: {}", stderr.trim()));
                }
            }
            Err(e) => errors.push(format!("pull outbox: {}", e)),
        }
    }

    // Push our outbox → remote inbox
    let outbox_dir = store.root().join("outbox");
    if outbox_dir.exists() {
        for entry in fs::read_dir(&outbox_dir)? {
            let entry = entry?;
            let fname = entry.file_name().to_string_lossy().to_string();
            if !fname.ends_with(".json") {
                continue;
            }
            if !fname.contains(&format!("_to-{}", peer.name)) && !fname.contains(&peer.id) {
                continue;
            }

            let src = entry.path().to_string_lossy().to_string();
            let dst = format!("{}:{}/inbox/{}", host, remote_path, fname);
            match rsync(&src, &dst).await {
                Ok(_) => {
                    archive_sent(&outbox_dir, &fname)?;
                    pushed.push(fname);
                }
                Err(e) => errors.push(format!("push {}: {}", entry.file_name().to_string_lossy(), e)),
            }
        }
    }

    fs::write(
        peer_dir.join(".last-sync"),
        chrono::Utc::now().to_rfc3339(),
    )?;

    Ok(SyncResult {
        peer_name: peer.name.clone(),
        pulled,
        pushed,
        errors,
    })
}

/// Archive delivered messages to outbox/.sent/ instead of deleting them.
/// Prevents duplicate delivery on next sync while preserving an audit trail.
/// Without this, every sync would re-push all outbox messages to the peer.
fn archive_sent(outbox_dir: &Path, fname: &str) -> Result<()> {
    let sent_dir = outbox_dir.join(".sent");
    fs::create_dir_all(&sent_dir)?;
    let src = outbox_dir.join(fname);
    let dst = sent_dir.join(fname);
    fs::rename(&src, &dst)?;
    Ok(())
}

async fn rsync(src: &str, dst: &str) -> Result<()> {
    let output = tokio::process::Command::new("rsync")
        .args(["-az", src, dst])
        .output()
        .await?;
    if !output.status.success() {
        anyhow::bail!("{}", String::from_utf8_lossy(&output.stderr).trim());
    }
    Ok(())
}

async fn rsync_dir(src: &str, dst: &str) -> Result<()> {
    let output = tokio::process::Command::new("rsync")
        .args(["-az", "--delete", src, dst])
        .output()
        .await?;
    if !output.status.success() {
        anyhow::bail!("{}", String::from_utf8_lossy(&output.stderr).trim());
    }
    Ok(())
}

File: src/sync.ts
// --- Transport design ---
// Two transports, one protocol. HTTP for WAN (daemon serves context over the internet),
// SSH/rsync for LAN (zero config if you already have SSH keys — uses ~/.ssh/config aliases
// so agents reference hostnames, never raw IPs that change). Both transports do the same
// thing: pull CONTEXT.md + PROFILE.md + shared/ + knowledge/, push outbox → peer inbox.

import { readFile, writeFile, mkdir, readdir, rename } from "node:fs/promises";
import { join, resolve } from "node:path";
import { existsSync } from "node:fs";
import { execFile as execFileCb } from "node:child_process";
import { promisify } from "node:util";
import dns from "node:dns/promises";
import { ContextStore } from "./store.js";

/** Block SSRF: reject URLs pointing to private/reserved IP ranges. */
export async function checkSsrf(url: string): Promise<void> {
  const parsed = new URL(url);
  const hostname = parsed.hostname.replace(/^\[|\]$/g, ""); // strip IPv6 brackets
  try {
    const { address } = await dns.lookup(hostname);
    const parts = address.split(".").map(Number);
    if (
      address === "127.0.0.1" || address === "::1" || address === "0.0.0.0" ||
      parts[0] === 10 ||
      (parts[0] === 172 && parts[1] >= 16 && parts[1] <= 31) ||
      (parts[0] === 192 && parts[1] === 168) ||
      (parts[0] === 169 && parts[1] === 254) ||
      address.startsWith("fc") || address.startsWith("fd") || address.startsWith("fe80")
    ) {
      throw new Error(`SSRF blocked: ${hostname} resolves to private address ${address}`);
    }
  } catch (e: any) {
    if (e.message.startsWith("SSRF blocked")) throw e;
    // DNS resolution failed — allow (could be .local or SSH alias)
  }
}

const execFile = promisify(execFileCb);

// Strip dangerous HTML from peer content before writing to disk.
// Peer-synced files get read by agents/LLMs — malicious HTML could execute
// if rendered in a browser or trick an LLM into acting on injected instructions.
function sanitizePeerContent(raw: string): string {
  return raw
    .replace(/<script[\s\S]*?<\/script>/gi, "[REMOVED: script tag]")
    .replace(/<iframe[\s\S]*?<\/iframe>/gi, "[REMOVED: iframe tag]")
    .replace(/<object[\s\S]*?<\/object>/gi, "[REMOVED: object tag]")
    .replace(/<embed[\s\S]*?>/gi, "[REMOVED: embed tag]")
    .replace(/<link[\s\S]*?>/gi, "[REMOVED: link tag]")
    .replace(/on\w+\s*=\s*["'][^"']*["']/gi, "[REMOVED: event handler]")
    .replace(/javascript\s*:/gi, "[REMOVED: javascript URI]");
}

export interface SyncResult {
  peerName: string;
  pulled: string[];
  pushed: string[];
  errors: string[];
}

interface Transport {
  type: "http" | "ssh";
  baseUrl?: string;
  host?: string;
  path?: string;
}

// Archive instead of delete: preserves audit trail and lets agents review what was sent.
// Without this, sync would re-deliver the same message every cycle.
// relPath can be "file.json" (flat, legacy) or "recipientDir/file.json" (new subdir layout).
async function archiveSent(outboxRoot: string, relPath: string): Promise<void> {
  // Path traversal defense: resolve and verify we stay under outboxRoot
  const fullPath = resolve(outboxRoot, relPath);
  if (!fullPath.startsWith(resolve(outboxRoot) + "/")) {
    throw new Error(`Path traversal blocked: ${relPath}`);
  }
  const dir = join(fullPath, "..");
  const fname = relPath.includes("/") ? relPath.split("/").pop()! : relPath;
  const sentDir = join(dir, ".sent");
  await mkdir(sentDir, { recursive: true });
  await rename(fullPath, join(sentDir, fname));
}

function parseUrl(url: string): Transport {
  if (url.startsWith("http://") || url.startsWith("https://")) {
    return { type: "http", baseUrl: url.replace(/\/$/, "") };
  } else if (url.startsWith("ssh://")) {
    const rest = url.slice(6);
    const colonIdx = rest.indexOf(":");
    if (colonIdx === -1) throw new Error("SSH URL must be ssh://host:/path");
    const host = rest.slice(0, colonIdx);
    const path = rest.slice(colonIdx + 1);
    // Prevent argument injection: rsync treats leading "-" as flags, and shell
    // metacharacters could escape the execFile boundary on some platforms.
    if (host.startsWith("-") || path.startsWith("-")) {
      throw new Error("Invalid SSH URL: host/path cannot start with '-'");
    }
    if (/[;|`$&(){}\s\n\r]/.test(host)) {
      throw new Error("Invalid SSH URL: host contains shell metacharacters");
    }
    if (/[;|`$&(){}\s\n\r]/.test(path)) {
      throw new Error("Invalid SSH URL: path contains shell metacharacters");
    }
    return { type: "ssh", host, path };
  }
  throw new Error(`Unknown URL scheme: ${url}. Use http:// or ssh://`);
}

/** Try to deliver a single outbox message immediately. Returns true if delivered.
 *  filename can be "recipientDir/msg.json" (new) or "flat.json" (legacy). */
export async function deliverOne(store: ContextStore, peerName: string, filename: string): Promise<boolean> {
  const config = await store.readConfig();
  const peer = config.peers.find((p) => p.name === peerName || p.id === peerName);
  if (!peer) return false;

  const outboxRoot = join(store.root, "outbox");
  const filePath = join(outboxRoot, filename);
  if (!existsSync(filePath)) return false;

  const baseName = filename.includes("/") ? filename.split("/").pop()! : filename;

  try {
    const transport = parseUrl(peer.url);

    if (transport.type === "http") {
      await checkSsrf(transport.baseUrl!);
      const body = await readFile(filePath, "utf-8");
      // Append ?to={name} for multi-tenant hosted mailboxes (inbox.openfused.dev).
      // Self-hosted daemons ignore the query param (single-tenant).
      const inboxUrl = `${transport.baseUrl}/inbox?to=${encodeURIComponent(peerName)}`;
      const r = await fetch(inboxUrl, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body,
      });
      if (!r.ok) return false;
    } else {
      await execFile("rsync", [
        "-az", filePath,
        `${transport.host}:${transport.path}/inbox/${baseName}`,
      ]);
    }

    // Delivered — archive to .sent/
    await archiveSent(outboxRoot, filename);
    return true;
  } catch {
    return false; // stays in outbox for next sync
  }
}

export async function syncAll(store: ContextStore): Promise<SyncResult[]> {
  const config = await store.readConfig();
  const results: SyncResult[] = [];
  for (const peer of config.peers) {
    try {
      results.push(await syncPeer(store, peer));
    } catch (e: any) {
      results.push({ peerName: peer.name, pulled: [], pushed: [], errors: [e.message] });
    }
  }
  return results;
}

export async function syncOne(store: ContextStore, peerName: string): Promise<SyncResult> {
  const config = await store.readConfig();
  const peer = config.peers.find((p) => p.name === peerName || p.id === peerName);
  if (!peer) throw new Error(`Peer not found: ${peerName}`);
  return syncPeer(store, peer);
}

async function syncPeer(
  store: ContextStore,
  peer: { name: string; id: string; url: string },
): Promise<SyncResult> {
  const transport = parseUrl(peer.url);
  const peerDir = join(store.root, ".peers", peer.name);
  await mkdir(peerDir, { recursive: true });

  if (transport.type === "http") {
    return syncHttp(store, peer, transport.baseUrl!, peerDir);
  } else {
    return syncSsh(store, peer, transport.host!, transport.path!, peerDir);
  }
}

// --- HTTP sync ---

async function syncHttp(
  store: ContextStore,
  peer: { name: string; id: string },
  baseUrl: string,
  peerDir: string,
): Promise<SyncResult> {
  const pulled: string[] = [];
  const pushed: string[] = [];
  const errors: string[] = [];

  // SSRF check: block requests to private/reserved IPs
  await checkSsrf(baseUrl);

  // Try /read/{file} (full mode) then /profile (public mode) for PROFILE.md.
  // CONTEXT.md 404s are silently skipped — peers in public mode don't serve it.
  for (const file of ["CONTEXT.md", "PROFILE.md"]) {
    try {
      const url = file === "PROFILE.md"
        ? `${baseUrl}/profile`  // public mode serves /profile not /read/PROFILE.md
        : `${baseUrl}/read/${file}`;
      let resp = await fetch(url);
      // Fallback: try /read/ path if /profile didn't work (full mode daemon)
      if (!resp.ok && file === "PROFILE.md") {
        resp = await fetch(`${baseUrl}/read/${file}`);
      }
      if (resp.ok) {
        const raw = sanitizePeerContent(await resp.text());
        const esc = (s: string) => s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
        const wrapped = `<external_content_unverified from="${esc(peer.name)}" file="${esc(file)}">\n${raw}\n</external_content_unverified>`;
        await writeFile(join(peerDir, file), wrapped);
        pulled.push(file);
      }
      // Don't report 404s as errors — peer may be in public mode
    } catch {}
  }

  for (const dir of ["shared", "knowledge"]) {
    try {
      const resp = await fetch(`${baseUrl}/ls/${dir}`);
      if (!resp.ok) continue;
      const files = (await resp.json()) as { name: string; is_dir: boolean }[];
      const localDir = join(peerDir, dir);
      await mkdir(localDir, { recursive: true });
      for (const f of files) {
        if (f.is_dir) continue;
        // Remote peer controls this filename — must sanitize before writing to local disk.
        // Basename extraction blocks "../../../etc/passwd" style traversal from a malicious peer.
        const safeName = f.name.split("/").pop()!.split("\\").pop()!;
        if (!safeName || safeName.includes("..")) continue;
        const r = await fetch(`${baseUrl}/read/${dir}/${safeName}`);
        if (r.ok) {
          const raw = sanitizePeerContent(Buffer.from(await r.arrayBuffer()).toString("utf-8"));
          const esc = (s: string) => s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
          const wrapped = `<external_content_unverified from="${esc(peer.name)}" file="${esc(safeName)}">\n${raw}\n</external_content_unverified>`;
          await writeFile(join(localDir, safeName), wrapped);
          pulled.push(`${dir}/${safeName}`);
        }
      }
    } catch (e: any) {
      errors.push(`${dir}/: ${e.message}`);
    }
  }

  // Pull peer's outbox for messages addressed to us (HTTP version).
  // Authenticated: we sign a challenge proving we own this name, so the daemon
  // only serves outbox to the actual recipient. Prevents metadata enumeration.
  const config = await store.readConfig();
  const myName = config.name;
  const inboxDir = join(store.root, "inbox");
  await mkdir(inboxDir, { recursive: true });
  try {
    const { signChallenge } = await import("./crypto.js");
    const timestamp = new Date().toISOString();
    const challenge = `OUTBOX:${myName}:${timestamp}`;
    const { signature, publicKey } = await signChallenge(store.root, challenge);

    const resp = await fetch(`${baseUrl}/outbox/${myName}`, {
      headers: {
        "X-OpenFuse-PublicKey": publicKey,
        "X-OpenFuse-Signature": signature,
        "X-OpenFuse-Timestamp": timestamp,
      },
    });
    if (resp.ok) {
      const messages = (await resp.json()) as any[];
      for (const msg of messages) {
        const ts = (msg.timestamp || new Date().toISOString()).replace(/[:.]/g, "-");
        const from = msg.from || "unknown";
        const safeFrom = from.replace(/[^a-zA-Z0-9\-_]/g, "");
        const safeTs = ts.replace(/[^a-zA-Z0-9\-_]/g, "");
        const fname = `${safeTs}_from-${safeFrom}_to-${myName}.json`;
        const outboxFile = msg._outboxFile; // filename on sender's outbox
        const dest = join(inboxDir, fname);
        if (!existsSync(dest)) {
          // Strip the _outboxFile metadata before saving
          const { _outboxFile, ...cleanMsg } = msg;
          await writeFile(dest, JSON.stringify(cleanMsg, null, 2));
          pulled.push(`outbox→${fname}`);

          // ACK: tell sender to move this message to .sent/
          if (outboxFile) {
            try {
              const ackTs = new Date().toISOString();
              const ackChallenge = `ACK:${myName}:${outboxFile}:${ackTs}`;
              const ackSig = await signChallenge(store.root, ackChallenge);
              await fetch(`${baseUrl}/outbox/${myName}/${outboxFile}`, {
                method: "DELETE",
                headers: {
                  "X-OpenFuse-PublicKey": ackSig.publicKey,
                  "X-OpenFuse-Signature": ackSig.signature,
                  "X-OpenFuse-Timestamp": ackTs,
                },
              });
            } catch {} // best-effort ACK
          }
        }
      }
    }
  } catch {}

  // Push outbox → peer inbox (scan subdirs named {peer}-{fp}/)
  const outboxDir = join(store.root, "outbox");
  if (existsSync(outboxDir)) {
    for (const entry of await readdir(outboxDir, { withFileTypes: true })) {
      // Match subdirs starting with peer name (new format: name-FINGERPRINT/)
      if (entry.isDirectory() && entry.name.startsWith(`${peer.name}-`)) {
        const subDir = join(outboxDir, entry.name);
        for (const fname of await readdir(subDir)) {
          if (!fname.endsWith(".json")) continue;
          const relPath = `${entry.name}/${fname}`;
          try {
            const body = await readFile(join(subDir, fname), "utf-8");
            const inboxUrl = `${baseUrl}/inbox?to=${encodeURIComponent(peer.name)}`;
            const r = await fetch(inboxUrl, {
              method: "POST",
              headers: { "Content-Type": "application/json" },
              body,
            });
            if (r.ok) {
              await archiveSent(outboxDir, relPath);
              pushed.push(relPath);
            } else errors.push(`push ${relPath}: HTTP ${r.status}`);
          } catch (e: any) {
            errors.push(`push ${relPath}: ${e.message}`);
          }
        }
      }
      // Legacy flat files (pre-subdir format)
      if (entry.isFile() && entry.name.endsWith(".json")) {
        if (!entry.name.includes(`_to-${peer.name}-`) && !entry.name.includes(`_to-${peer.name}.json`)) continue;
        try {
          const body = await readFile(join(outboxDir, entry.name), "utf-8");
          const inboxUrl = `${baseUrl}/inbox?to=${encodeURIComponent(peer.name)}`;
          const r = await fetch(inboxUrl, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body,
          });
          if (r.ok) {
            await archiveSent(outboxDir, entry.name);
            pushed.push(entry.name);
          } else errors.push(`push ${entry.name}: HTTP ${r.status}`);
        } catch (e: any) {
          errors.push(`push ${entry.name}: ${e.message}`);
        }
      }
    }
  }

  await writeFile(join(peerDir, ".last-sync"), new Date().toISOString());
  return { peerName: peer.name, pulled, pushed, errors };
}

// --- SSH sync (rsync) ---

async function syncSsh(
  store: ContextStore,
  peer: { name: string; id: string },
  host: string,
  remotePath: string,
  peerDir: string,
): Promise<SyncResult> {
  const pulled: string[] = [];
  const pushed: string[] = [];
  const errors: string[] = [];

  for (const file of ["CONTEXT.md", "PROFILE.md"]) {
    try {
      await execFile("rsync", ["-az", `${host}:${remotePath}/${file}`, join(peerDir, file)]);
      pulled.push(file);
    } catch (e: any) {
      errors.push(`${file}: ${e.stderr || e.message}`);
    }
  }

  for (const dir of ["shared", "knowledge"]) {
    const localDir = join(peerDir, dir);
    await mkdir(localDir, { recursive: true });
    try {
      await execFile("rsync", ["-az", "--delete", `${host}:${remotePath}/${dir}/`, `${localDir}/`]);
      pulled.push(`${dir}/`);
    } catch (e: any) {
      errors.push(`${dir}/: ${e.stderr || e.message}`);
    }
  }

  // Pull peer's outbox for messages addressed to us — peer may be behind NAT
  // and can't push to us, so we grab messages they left in their outbox for us.
  // New layout: outbox/{name}-{fp}/*.json — pull from all dirs starting with our name.
  const config = await store.readConfig();
  const myName = config.name;
  const inboxDir = join(store.root, "inbox");
  await mkdir(inboxDir, { recursive: true });

  // New subdir format: pull outbox/{myName}-*/*.json into a temp dir (preserves structure),
  // then move the .json files into inbox/ (flattened). rsync --include handles the filtering;
  // we avoid ssh commands to prevent shell injection via host/path values.
  const tmpPull = join(store.root, ".tmp-outbox-pull");
  try {
    await mkdir(tmpPull, { recursive: true });
    await execFile("rsync", [
      "-az", "--ignore-existing",
      "--include", `${myName}-*/`,
      "--include", `${myName}-*/*.json`,
      "--exclude", "*",
      `${host}:${remotePath}/outbox/`,
      `${tmpPull}/`,
    ]);
    // Flatten: move .json files from subdirs into inbox/
    if (existsSync(tmpPull)) {
      for (const subEntry of await readdir(tmpPull, { withFileTypes: true })) {
        if (!subEntry.isDirectory() || !subEntry.name.startsWith(`${myName}-`)) continue;
        const subPath = join(tmpPull, subEntry.name);
        for (const fname of await readdir(subPath)) {
          if (!fname.endsWith(".json")) continue;
          const dest = join(inboxDir, fname);
          if (!existsSync(dest)) {
            await rename(join(subPath, fname), dest);
          }
        }
      }
    }
    pulled.push("outbox→inbox");
  } catch (e: any) {
    if (!String(e.stderr || e.message).includes("No such file")) {
      errors.push(`pull outbox (subdir): ${e.stderr || e.message}`);
    }
  } finally {
    // Clean up temp dir
    try { await (await import("node:fs/promises")).rm(tmpPull, { recursive: true, force: true }); } catch {}
  }

  // Legacy flat format: outbox/*_to-{name}*.json
  try {
    await execFile("rsync", [
      "-az", "--ignore-existing",
      "--include", `*_to-${myName}-*.json`,
      "--include", `*_to-${myName}.json`,
      "--include", `*_to-all.json`,
      "--exclude", "*",
      `${host}:${remotePath}/outbox/`,
      `${inboxDir}/`,
    ]);
  } catch (e: any) {
    if (!String(e.stderr || e.message).includes("No such file")) {
      errors.push(`pull outbox (legacy): ${e.stderr || e.message}`);
    }
  }

  // Push our outbox → peer inbox (scan subdirs named {peer}-{fp}/)
  const outboxDir = join(store.root, "outbox");
  if (existsSync(outboxDir)) {
    for (const entry of await readdir(outboxDir, { withFileTypes: true })) {
      if (entry.isDirectory() && entry.name.startsWith(`${peer.name}-`)) {
        const subDir = join(outboxDir, entry.name);
        for (const fname of await readdir(subDir)) {
          if (!fname.endsWith(".json")) continue;
          const relPath = `${entry.name}/${fname}`;
          try {
            await execFile("rsync", ["-az", join(subDir, fname), `${host}:${remotePath}/inbox/${fname}`]);
            await archiveSent(outboxDir, relPath);
            pushed.push(relPath);
          } catch (e: any) {
            errors.push(`push ${relPath}: ${e.stderr || e.message}`);
          }
        }
      }
      // Legacy flat files
      if (entry.isFile() && entry.name.endsWith(".json")) {
        if (!entry.name.includes(`_to-${peer.name}-`) && !entry.name.includes(`_to-${peer.name}.json`)) continue;
        try {
          await execFile("rsync", ["-az", join(outboxDir, entry.name), `${host}:${remotePath}/inbox/${entry.name}`]);
          await archiveSent(outboxDir, entry.name);
          pushed.push(entry.name);
        } catch (e: any) {
          errors.push(`push ${entry.name}: ${e.stderr || e.message}`);
        }
      }
    }
  }

  await writeFile(join(peerDir, ".last-sync"), new Date().toISOString());
  return { peerName: peer.name, pulled, pushed, errors };
}
